### PR TITLE
Tcp interfacing

### PR DESCRIPTION
**Pull request by dstockhouse**
Review system/src/tcp.c and system/test/tcptest.c for TCP interface abstraction, as well as placement of new files in the system/ directory. All other files there have been around since before the project started, so if you feel up to it feel free to review. I'm working on unit tests for the other system modules that haven't yet been tested

### CODE REVIEW

File: navigation/test/vn200_test.c [hunk 1]
```diff
@@ -8,6 +8,6 @@ BeforeEach(VN200) {}
 AfterEach(VN200) {}
 
 Ensure(VN200, dummy_test) {
-	assert_that(1 == 1);
+    assert_that(1, is_equal_to(1));
```
Comment (seanybaggins) [inline navigation/test/vn200_test.c]: Can we delete this dummy test?

File: system/inc/config.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+#define L_INFO      2
+
+#define CONTROL_DEBUG_L_MASK    L_DEBUG
+
+#define DEBUG_OUT_PRINTF
+#undef DEBUG_OUT_SYSLOG
+#undef DEBUG_OUT_LOGFILE
+
+
+// Placeholders. Fill in with real data during test and integration
```
Comment (seanybaggins) [inline system/inc/config.h:35]: Correct IP addresses for controls and guidance is documented in the lucidcharts/mr_fusion_capstone/interphasing

File: system/inc/config.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+#define GUIDANCE_IP_ADDR    "XXX.XXX.XXX.XXX"
```
Comment (seanybaggins) [inline system/inc/config.h]: 192.168.1.1

File: system/inc/config.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+#define NAVIGATION_IP_ADDR  "XXX.XXX.XXX.XXX"
+#define CONTROL_IP_ADDR     "XXX.XXX.XXX.XXX"
```
Comment (seanybaggins) [inline system/inc/config.h]: 192.168.1.2

File: system/inc/config.h [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+#define DRAGONBOARD_IP_ADDR "XXX.XXX.XXX.XXX"
+
+// Need individual port for each pair of devices
+#define GUIDANCE_TCP_PORT       38000
+#define NAVIGATION_TCP_PORT     38010
```
Comment (seanybaggins) [inline system/inc/config.h]: 31402

File: system/inc/config.h [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+#define CONTROL_TCP_PORT        38020
```
Comment (seanybaggins) [inline system/inc/config.h]: 31401

File: system/inc/config.h [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+#define DRAGONBOARD_TCP_PORT    38030
```
Comment (seanybaggins) [inline system/inc/config.h]: 31403

File: system/test/tcptest.c [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+#include <cgreen/cgreen.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <unistd.h>
+#include <poll.h>
+
+#include "tcp.h"
+
+#define IP_ADDR     "127.0.0.1"
```
Comment (seanybaggins) [inline system/test/tcptest.c:28]: Do these unit tests become subject to race conditions when they are all in a separate process and communicate over the same localhoast?

File: system/src/tcp.c [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+    struct sockaddr_in socketAddress;
+
+    // Exit on error if invalid pointer
+    if (ipAddr == NULL) {
+        return -1;
+    }
+
+    // Configure socket address
+    socketAddress.sin_family = AF_INET;
+    socketAddress.sin_port = htons(port);
```
Comment (seanybaggins) [inline system/src/tcp.c:106]: Still not 100% sure after looking at the comments where htons is defined what this is doing.

File: system/test/tcptest.c [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+Ensure(TCPInterface, client_fails_with_async_connect) {
+
+    client_fd = TCPClientInit();
+    assert_that(client_fd, is_not_equal_to(-1));
+
+    // Attempt to connect to a nonexistent server
+    rc = TCPClientTryConnect(client_fd, IP_ADDR, SERVER_PORT);
+    assert_that(rc, is_equal_to(-1));
+    assert_that(errno, is_equal_to(ECONNREFUSED));
+    // perror("TCPClientTryConnect failed (expected)");
```
Comment (seanybaggins) [inline system/test/tcptest.c]: Remove comment

File: system/test/tcptest.c [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+
+    // Set nonblocking for attempted accept
+    rc = TCPSetNonBlocking(server_fd);
+    assert_that(rc, is_not_equal_to(-1));
+
+    // Attempt to accept a nonexistent client
+    rc = TCPServerTryAccept(server_fd);
+    assert_that(rc, is_equal_to(-1));
+    assert_that(errno, is_equal_to(EAGAIN));
+    // perror("TCPServerTryAccept failed (expected)");
```
Comment (seanybaggins) [inline system/test/tcptest.c]: remove comment

File: system/test/tcptest.c [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+
+    // Open to be closed
+    client_fd = TCPClientInit();
+
+}
+
+Ensure(TCPInterface, client_server_connect_loopback) {
+
+    setup_sockets();
```
Comment (seanybaggins) [inline system/test/tcptest.c]: Should this unit test have an assertion?

**Review by seanybaggins — CHANGES_REQUESTED** [review]
The code looks very clean. Could you update our make file and Travis to include coverage reports for our c code? This would give me a better sense of what code has been tested and what has not. I can get on the phone if we want to tag team this one since it appears to be a mix of our skills. If you decide to tackle it yourself, here is a link to get you started for adding coverage reports for our c code.

https://github.com/codecov/example-c

File: system/test/tcptest.c (hunk 8 above)
Comment (dstockhouse) [reply system/test/tcptest.c:28]: From the cgreen doc section 2.8: "Just before calling the BeforeEach() (or setup) function, Cgreen fork():s. The main process waits for the test to complete normally or die."
Each test runs in its own process, but they do not run in parallel so there is no contention for the same resources

File: system/test/tcptest.c (hunk 12 above)
Comment (dstockhouse) [reply system/test/tcptest.c]: The function setup_sockets() has assertions built in, but I made it better and more clear in the next rev

**Review by rockwec1 — APPROVED** [review]

**Review by seanybaggins — APPROVED** [review]

**Pull request merged**